active/0004-extensibility-support-for-deployment-stacks.md

```diff
@@ -0,0 +1,405 @@
+{
+   // ... deployment contract
+   "extensions": [
+      {
+         "name": "Kubernetes",
+         "alias": "k8s",
+         "version": "1.0.0",
+         "deploymentId": "/subscriptions/.../resourceGroups/.../providers/Microsoft.Resources/deployments/...",
+         "config": {
+            "namespace": {
+               "type": "string",
+               "value": "default"
+            },
+            "kubeConfig": {
+               "type": "directive",
+               "evaluation": [
+                  {
+                     "type": "ArmApiCall",
+                     "resourceId": "/subscriptions/.../resourceGroups/.../Microsoft.ContainerService/managedClusters/...",
+                     "method": "GET",
+                     "apiVersion": "2024-02-01",
+                     "apiAction": "/listClusterAdminCredential",
+                     "query": "?a=1&b=2" // optional
+                     "body": { } // optional
+                  },
+                  {
+                     "type": "JSONPath",
+                     "path": "kubeconfig[0].value" // decompose?
+                  }
+               ]
+            }
+         }
+      }
+   ]
+}
+```
+
+With the deployments service returning a list of all extensions (across all nested deployments) with their configuration
+directives and returning all extension resources deployed in the template (across all nested deployments), the stacks 
+service can process each extension resource and use the "deploymentId" and "extensionAlias" properties as a composite 
+key to the data in the extensions array.
+
+## Drawbacks
+
+The biggest drawback of this approach is that extension configuration data cannot be sourced from deployment parameters.
+This design decision is driven solely to support the stacks user experience not needing to prompt for the parameter 
+values of the last stack deployment as user secrets cannot be persisted long term. This has the side effect of disallowing
+the use of non-secret parameters as well, such as the "namespace" in the Kubernetes extension use case. There could be
+a constraint to allow only non-secure parameter usage in extension configurations, but that could end up encouraging the
+user to work around that by not using secure parameters for data that should be secured. Alternatively, the secret portion
+of the extension configuration could be split into from non-secret data. The secret portion would adopt the design proposed
+in this REP and the non-secret portion could be kept as-is, but there would need to be a way for the extension to
+provide metadata about the properties of the configuration schema.
+
+The move of extension configuration definitions to the parent deployment properties along with the language expression 
+constraints it imposes means that extension resources that require a configuration will no longer be deployable
+in a root deployment. There is no mechanism at the root level to provide this data. Client side changes would need to
```

Comment:
If the `extensions` property is accepted on a nested deployment resource, that means it is part of the deployments PUT contract. Is the blocker here just adding new options to the CLI?

Reply:
IMHO a temporary workaround could be to continue supporting the use of parameters to specify extension configurations within the module itself for the time being. We can decide later whether to introduce new CLI options or solely rely on .bicepdeploy files for extension configurations.

Reply:
Although that creates some challenge as Kyle highlighted below in the alternatives section. A less ideal workaround is to ask customers to always declare a module to provide extension configs.

Reply:
Yeah the main blocker is a mechanism on frontends to provide extension configs but only in formats that can be compiled to the format expected in the PUT contract. It's like a parameters file but without accepting dynamic inputs (inline parameters in the CLI).

Reply:
I updated the design to propose changes to the parameters file to support extension configs.